src/Microsoft.AspNetCore.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -79,6 +79,7 @@ public MediaType(string mediaType, int offset, int? length)
             var subTypeLength = GetSubtypeLength(mediaType, offset + typeLength, out subType);
             if (subTypeLength == 0)
             {
+                // ??? Should this case also reset Type? A type without a subtype is invalid.
```

Comment:
I realize that this type needs some "love", in general, consumers of this type should use methods provided by the type like "IsSubsetOf" and these methods should be able to correctly handle incorrect input. Having an instance of MediaType doesn't mean you have a valid MediaType, it can be default(MediaType) or it might be a media type with valid type but invalid subtype, valid type and subtype and invalid parameter.

I would say that we should do the following (separate bug / separate PR). Go through the algorithms in this type (GetParameter, IsSubsetOf, etc) and make sure they return the appropriate result against invalid input (null and false for the previous cases). (For that, we need to iterate to the end of the parameter set or consider "recovery strategies" for relaxed parsing rules of our header (given the amount of invalid clients out there)

We should add more tests to cover scenarios with invalid media types and we should make sure that the q parameter is handled appropriately.

Finally, we should consider adding an IsValid computed property on MediaType that checks that Type != default(StringSegment), SubType != default(StringSegment) and that iterates through all the parameters and makes sure that they don't fail.

The goal of this type is to not allocate when possible and to be as lazy as possible as it's on a performance critical path. If we need to add more fields to this struct, (for example to cache the validity of a media type) we should measure the performance impact on the type.